azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages_mqtt.py

```diff
@@ -58,7 +58,27 @@ def on_connected():
                 self.complete_op(op)
 
             self.provider.on_mqtt_connected = on_connected
-            self.provider.connect(self.sas_token)
+            try:
+                self.provider.connect(self.sas_token)
+            except BaseException:
```

Comment:
>BaseException [](start = 19, length = 13)

is this exception only possible during connect reconnect and disconnect ? and NOT during publish or any thing else ?
if so may be we need to update docs in the provider functions for connect , disconnect and reconnect saying it may raise exceptions .

Reply:
Excepting on `BaseException` seems dangerous since it catches all kinds of stuff that you probably don't want (KeyboardInterrupt).

Also, in looking this up it turns out all the places we've been using bare `except` statements are implicitly doing this... so that's also probably a problem in those places. Perhaps instead of addressing this now, we can address it with a future PR that improves the exception handling story, and for now just use a bare `except` for consistency (which does the same thing as `except BaseException`)

---
In reply to: [285815533](https://github.com/Azure/azure-iot-sdk-python-preview/pull/94#discussion_r285815533) [](ancestors = 285815533)

Reply:
It's not that the functions _can_ raise exceptions.  _Any_ function can raise an exception.  This is here because we need to something _special_ if an exception is raised.  I added this comment to the code.  Hopefully it explains it well.

A note on exceptions handling in Connect, Disconnct, and Reconnet:
 
All calls into self.provider can raise an exception, and this is OK.
The exception handler in PipelineStage.run_op() will catch these errors
and propagate them to the caller.  This is an intentional design of the
pipeline, that stages, etc, don't need to worry about catching exceptions
except for special cases.

The code right below this comment is This is a special case.  In addition
to this "normal" exception handling, we add another exception handler
into this class' Connect, Reconnect, and Disconnect code.  We need to
do this because provider.on_mqtt_connected and provider.on_mqtt_disconnected
are both _handler_ functions instead of _callbacks_.

Because they're handlers instead of callbacks, we need to change the
handlers while the connection is established.  We do this so we can
know when the provider is connected so we can move on to the next step.
Once the connection is established, we change the handler back to its
old value before finishing.

The exception handling below is to reset the handler back to its original
value in the case where provider.connect raises an exception.  Again,
this extra exception handling is only necessary in the Connect, Disconnect,
and Reconnect case because they're the only cases that use handlers instead
of callbacks.


---
In reply to: [287157069](https://github.com/Azure/azure-iot-sdk-python-preview/pull/94#discussion_r287157069) [](ancestors = 287157069,285815533)